### PR TITLE
`RuntimeGenesiConfig`: json macro added

### LINKED ISSUE #5700: Only customized fields in preset json blob?
Probably not super important, but wanted to know your opinion on that, so I am leaving it here for your consideration.

This is about how to provide preset's JSON and what is potential outcome.

There are two approaches.

<details><summary>JSON blob, containing only overridden fields</summary>

```
fn preset() -> serde_json::Value {
    serde_json::json!({
        "balances": BalancesConfig {
            balances: endowed_accounts.iter().cloned().map(|k| (k, 1u128 << 60)).collect::<Vec<_>>(),
        },
        "parachainInfo": ParachainInfoConfig {
            parachain_id: id,
            ..Default::default()
        },
        "collatorSelection": CollatorSelectionConfig {
            invulnerables: invulnerables.iter().cloned().map(|(acc, _)| acc).collect(),
            candidacy_bond: BRIDGE_HUB_ROCOCO_ED * 16,
            ..Default::default()
        },
        "session": SessionConfig {
            keys: invulnerables
                .into_iter()
                .map(|(acc, aura)| {
                    (
                        acc.clone(),                         // account id
                        acc,                                 // validator id
                        SessionKeys { aura },            // session keys
                    )
                })
                .collect(),
            ..Default::default()
        },
        "polkadotXcm": PolkadotXcmConfig {
            safe_xcm_version: Some(SAFE_XCM_VERSION),
            ..Default::default()
        },
        "bridgeWestendGrandpa": BridgeWestendGrandpaConfig {
            owner: bridges_pallet_owner.clone(),
            ..Default::default()
        },
        "bridgeWestendMessages": BridgeWestendMessagesConfig {
            owner: bridges_pallet_owner.clone(),
            ..Default::default()
        },
        "ethereumSystem": EthereumSystemConfig {
            para_id: id,
            asset_hub_para_id,
            ..Default::default()
        }
    })
}
```
</details>

and second:
<details><summary>full `RuntimeGenesisConfig` blob, containing defaults and overwritten fields:</summary>

```
fn preset() -> serde_json::Value {
    let config = RuntimeGenesisConfig {
        balances: BalancesConfig {
            balances: endowed_accounts
                .iter()
                .cloned()
                .map(|k| (k, 1u128 << 60))
                .collect::<Vec<_>>(),
        },
        parachain_info: ParachainInfoConfig { parachain_id: id, ..Default::default() },
        collator_selection: CollatorSelectionConfig {
            invulnerables: invulnerables.iter().cloned().map(|(acc, _)| acc).collect(),
            candidacy_bond: BRIDGE_HUB_ROCOCO_ED * 16,
            ..Default::default()
        },
        session: SessionConfig {
            keys: invulnerables
                .into_iter()
                .map(|(acc, aura)| {
                    (
                        acc.clone(),          // account id
                        acc,                  // validator id
                        SessionKeys { aura }, // session keys
                    )
                })
                .collect(),
            ..Default::default()
        },
        polkadot_xcm: PolkadotXcmConfig {
            safe_xcm_version: Some(SAFE_XCM_VERSION),
            ..Default::default()
        },
        bridge_westend_grandpa: BridgeWestendGrandpaConfig {
            owner: bridges_pallet_owner.clone(),
            ..Default::default()
        },
        bridge_westend_messages: BridgeWestendMessagesConfig {
            owner: bridges_pallet_owner.clone(),
            ..Default::default()
        },
        ethereum_system: EthereumSystemConfig {
            para_id: id,
            asset_hub_para_id,
            ..Default::default()
        },
        ..Default::default()
    };

    serde_json::to_value(config).expect("Could not build genesis config.")
}
```
</details>

When we generate two two runtime wasm blobs using two above approaches, and then we use `chain-spec-builder display-preset` util to get the json preset representation, and we display the diff, we will get following:
```
$ diff f.json j.json
2,5d1
<   "aura": {
<     "authorities": []
<   },
<   "auraExt": {},
58,68d53
<   "bridgePolkadotBulletinGrandpa": {
<     "initData": null,
<     "owner": null
<   },
<   "bridgePolkadotBulletinMessages": {
<     "openedLanes": [],
<     "operatingMode": {
<       "Basic": "Normal"
<     },
<     "owner": null
<   },
80,83d64
<   "bridgeWestendParachains": {
<     "operatingMode": "Normal",
<     "owner": null
<   },
99d79
<   "parachainSystem": {},
121,130d100
<   },
<   "system": {},
<   "transactionPayment": {
<     "multiplier": "1000000000000000000"
<   },
<   "xcmOverBridgeHubWestend": {
<     "openedBridges": []
<   },
<   "xcmOverPolkadotBulletin": {
<     "openedBridges": []
```

So the preset constructed from the full `RuntimeGenesisConfig` struct contains some fields that are just defaults and were not actually customized.

In practice, in most cases that should not matter, but it potentially may lead to some strange behavior - e.g. imagine fetching the preset to the file, then using it to against different runtime version (e.g. newer version with `transactionPayment::multiplier` updated) where defaults are changed may lead to some strange situation.

On the other hand - using JSON approach may lead to invalid JSON (and we would addition coverage in tests).

tagging @kianenigma , @bkchr.

_Originally posted by @michalkucharczyk in https://github.com/paritytech/polkadot-sdk/issues/5327#issuecomment-2343335131_

### PR DESCRIPTION
**Pull request by michalkucharczyk**
This PR adds `build_struct_json_patch` which helps to generate a JSON used for preset.

Here is doc and example:
https://github.com/paritytech/polkadot-sdk/blob/d868b858758d3886d16c8ba8feb3c93c188044f3/substrate/frame/support/src/generate_genesis_config.rs#L168-L266

And real-world usage:
https://github.com/paritytech/polkadot-sdk/blob/d868b858758d3886d16c8ba8feb3c93c188044f3/cumulus/parachains/runtimes/assets/asset-hub-rococo/src/genesis_config_presets.rs#L37-L61

Closes #5700

### CODE REVIEW

File: substrate/frame/support/src/genesis_builder_helper.rs [hunk 1]
```diff
@@ -55,3 +55,82 @@ where
...
+#[macro_export]
+macro_rules! runtime_genesis_config_json {
+	({ $( $key:ident : $value:expr ),* $(,)? }) => {{
+        let config = RuntimeGenesisConfig {
+            $( $key : $value, )*
+            ..Default::default()
+        };
+
+		#[inline]
+        fn compare_keys(
```
Comment (michalkucharczyk) [inline substrate/frame/support/src/genesis_builder_helper.rs:100]: We could also use [`convert_case`](https://docs.rs/convert_case/0.6.0/convert_case/) crate. But this probably would be unnecessary overhead.

File: substrate/frame/support/src/genesis_builder_helper.rs [hunk 2]
```diff
@@ -55,3 +55,82 @@ where
...
+///     runtime_genesis_config_json! ({a_field:31, b_field:127}),
+///     serde_json::json!({"aField":31, "bField":127})
+/// );
+/// assert_eq!(
+///     runtime_genesis_config_json! ({a_field:31}),
+///     serde_json::json!({"aField":31})
+/// );
+/// ```
+#[macro_export]
+macro_rules! runtime_genesis_config_json {
```
Comment (skunert) [inline substrate/frame/support/src/genesis_builder_helper.rs:92]: Not sure if this is needed, but we could allow to optionally specify the name of the struct to use. Such that one could to both `genesis_config_json!({a_field: 31, b_field: 127})` and `genesis_config_json!(CustomGenesisConfig, {x_field: 31, y_field: 127})`. Not sure if needed, but would make it more flexible. But have no strong feelings about it.

**Review by skunert — APPROVED** [review]

File: substrate/frame/support/src/genesis_builder_helper.rs (hunk 2 above)
Comment (skunert) [reply substrate/frame/support/src/genesis_builder_helper.rs:92]: Or if you manually want to use a different name for whatever reason.  e.g. use `crate::GenesisRuntimeConfig as CustomConfig`. But yeah maybe too niche. If someone wants this, we could add later.

File: substrate/frame/support/src/genesis_builder_helper.rs (hunk 1 above)
Comment (skunert) [reply substrate/frame/support/src/genesis_builder_helper.rs:100]: Thinking about it, does it handle all the weird edge cases? In theory, the pallet authors can name their field names whatever they like. Like `A___B`, `ABC_ABC` or whatever.  Maybe better to use the crate after all.

Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs:100]: Not sure if this is needed. The names in `RuntimeGenesisConfig` are also [generated](https://github.com/paritytech/polkadot-sdk/blob/f6d08e637694729df7cddb166954ee90c2da3ce1/substrate/frame/support/procedural/src/construct_runtime/expand/config.rs#L41) by the `construct_runtime` macro. I assume no weird naming will happen there - don't know if you can name your pallet with `_` within the name.


Anyway, I see that [Inflector](https://docs.rs/Inflector/latest/inflector/cases/camelcase/index.html) is already used in our codebase so I will rework this.
If something will pop-up we can fix it later.

Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs:100]: The other question is how serde_json is generating camel_case. We should use exactly the same implementation if we want to be 100% accurate.

Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs:100]: One more thought: `Inflector` is used in `procedural` and not in runtime.

File: substrate/frame/support/src/genesis_builder_helper.rs [hunk 3]
```diff
@@ -55,3 +55,82 @@ where
...
```
Comment (bkchr) [inline substrate/frame/support/src/genesis_builder_helper.rs:92]: But what about sub maps? Basically like:

```
pallet_whatever: { 
   i_only_want_to_provide_this_field: "cool",
}
```

Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs:92]: yeah, this proposal does not cover this. support for recursive partial structs would a bit more complex.

Comment (bkchr) [reply substrate/frame/support/src/genesis_builder_helper.rs:92]: Then we should support this as well.

File: substrate/frame/support/src/genesis_builder_helper.rs (hunk 1 above)
Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs]: sadly `Inflector` has no support for `nostd`.

File: substrate/frame/support/src/genesis_builder_helper.rs (hunk 2 above)
Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs]: done in: 6db4b41

File: substrate/frame/support/src/genesis_builder_helper.rs (hunk 1 above)
Comment (michalkucharczyk) [reply substrate/frame/support/src/genesis_builder_helper.rs]: > Thinking about it, does it handle all the weird edge cases? In theory, the pallet authors can name their field names whatever they like. Like `A___B`, `ABC_ABC` or whatever. Maybe better to use the crate after all.

This is done in 6db4b41

**Comment by michalkucharczyk** [comment]
I'll do one more renaming round.
Thinking about something like `init_partial...` which better reflects what macro actually does.

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 4]
```diff
@@ -0,0 +1,875 @@
...
+				);
+				$type {
+					$keyi:$value,
+					..Default::default()
+				}
+			},
+			..Default::default()
+		}
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident:  $type:tt { $($body:tt)* } ) => {
```
Comment (skunert) [inline substrate/frame/support/src/generate_genesis_config.rs:246]:
```suggestion
	($struct_type:ident, $all_keys:ident  @  $key:ident:  $type:ident { $($body:tt)* } ) => {
```
I think type can only be a type identifier here.

**Comment by michalkucharczyk** [comment]
> One thing I noticed while playing that is not supported is initilializing with full path, e.g. this fails:
>  ...


Should be fixed with d6c88547b15

**Review by skunert — APPROVED** [review]

File: substrate/frame/support/src/generate_genesis_config.rs (hunk 4 above)
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: yeah, right it can be identifier.

Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: done in: 0c24c455

**Review by bkchr — APPROVED** [review]
Machine! :D 

Nice!

I left some comments around macro hygiene, for the macro itself you have enough tests.

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 5]
```diff
@@ -0,0 +1,875 @@
...
+/// values. The macro then serializes the fully initialized structure into a JSON blob, retaining
+/// only the fields that were explicitly provided, either partially or fully initialized.
+///
+/// Using this macro prevents errors from manually creating JSON objects, such as typos or
+/// inconsistencies with the `struct_type` structure, by relying on the actual
+/// struct definition. This ensures the generated JSON is valid and reflects any future changes
+/// to the structure.
+///
+/// This macro assumes that the `serde(rename_all="camelCase")` attribute is applied to
+/// `RuntimeGenesisConfig`, which is typical for frame-based runtimes.
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: remove that piece in [0c24c45]

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 6]
```diff
@@ -0,0 +1,875 @@
...
+			loop {
+				match (ident_chars.find(|&c| c != '_'), camel_chars.next()) {
+					(None, None) => return true,
+					(None, Some(_)) | (Some(_), None) => return false,
+					(Some(i), Some(c)) =>
+						if i.to_ascii_uppercase() != c.to_ascii_uppercase() {
+							return false;
+						},
+				};
+			}
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: :see_no_evil: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 7]
```diff
@@ -0,0 +1,875 @@
...
+/// ```ignore
+/// RuntimeGenesisConfig {
+/// 		b_field {
+/// 			i_field: 2,
+/// 			..Default::default()
+/// 		},
+/// 		..Default::default()
+/// }
+/// ```
+/// while all other fields are initialized with default values. The macro serializes this, retaining
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+/// only the provided fields.
+#[macro_export]
+macro_rules! build_struct_json_patch {
+	(
+		$struct_type:ident { $($tail:tt)* }
+	) => {
+		{
+			use $crate::generate_genesis_config::{InitializedField, retain_initialized_fields};
+			extern crate alloc;
+			use alloc::{string::ToString, vec::Vec };
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 9]
```diff
@@ -0,0 +1,875 @@
...
+			let mut keys : Vec<InitializedField> = vec![];
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 10]
```diff
@@ -0,0 +1,875 @@
...
+/// ```
+/// while all other fields are initialized with default values. The macro serializes this, retaining
+/// only the provided fields.
+#[macro_export]
+macro_rules! build_struct_json_patch {
+	(
+		$struct_type:ident { $($tail:tt)* }
+	) => {
+		{
+			use $crate::generate_genesis_config::{InitializedField, retain_initialized_fields};
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 11]
```diff
@@ -0,0 +1,875 @@
...
+			extern crate alloc;
+			use alloc::{string::ToString, vec::Vec };
+			let mut keys : Vec<InitializedField> = vec![];
+			#[allow(clippy::needless_update)]
+			let struct_instance = build_struct_json_patch!($struct_type, keys @  { $($tail)* });
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 12]
```diff
@@ -0,0 +1,875 @@
...
+			let mut json_value =
+				serde_json::to_value(struct_instance).expect("serialization to json should work. qed");
+			retain_initialized_fields(&mut json_value, &keys, Default::default());
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 13]
```diff
@@ -0,0 +1,875 @@
...
+		$struct_type {
+			..build_struct_json_patch!($struct_type, $all_keys @ $($tail)*)
+		}
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident: $type:tt { $keyi:ident : $value:tt }  ) => {
+		$struct_type {
+			$key: {
+				$all_keys.push(InitializedField::Partial(stringify!($key).to_string()));
+				$all_keys.push(
+					InitializedField::Full(alloc::format!("{}.{}", stringify!($key), stringify!($keyi)))
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 14]
```diff
@@ -0,0 +1,875 @@
...
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident:  $type:tt { $($body:tt)* } ) => {
+		$struct_type {
+			$key: {
+				$all_keys.push(InitializedField::Partial(stringify!($key).to_string()));
+				let mut inner_keys = Vec::<InitializedField>::default();
+				let value = build_struct_json_patch!($type, inner_keys @ { $($body)* });
+				for i in inner_keys.iter_mut() {
+					i.add_prefix(stringify!($key));
+				};
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 15]
```diff
@@ -0,0 +1,875 @@
...
+/// This enum provides information about how it was initialized and the field name (as a `String`).
+///
+/// Intended to be used in `build_struct_json_patch` macro.
+#[derive(Debug)]
+pub enum InitializedField {
+	/// The field was partially initialized (e.g., specific fields within the struct were set
+	/// manually).
+	Partial(String),
+	/// The field was fully initialized (e.g., using `new()` or `default()` like methods).
+	Full(String),
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 16]
```diff
@@ -0,0 +1,875 @@
...
+			},
+			..Default::default()
+		}
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident:  $type:tt { $($body:tt)* },  $($tail:tt)*  ) => {
+		$struct_type {
+			$key : {
+				$all_keys.push(InitializedField::Partial(stringify!($key).to_string()));
+				let mut inner_keys = Vec::<InitializedField>::default();
+				let value = build_struct_json_patch!($type, inner_keys @ { $($body)* });
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 17]
```diff
@@ -0,0 +1,875 @@
...
+			.. build_struct_json_patch!($struct_type, $all_keys @ $($tail)*)
+		}
+	};
+	($struct_type:ident, $all_keys:ident  @  $key:ident: $value:expr, $($tail:tt)* ) => {
+		$struct_type {
+			$key: {
+				$all_keys.push(InitializedField::Full(stringify!($key).to_string()));
+				$value
+			},
+			..build_struct_json_patch!($struct_type, $all_keys @ $($tail)*)
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 18]
```diff
@@ -0,0 +1,875 @@
...
+				$all_keys.push(InitializedField::Partial(stringify!($key).to_string()));
+				let mut inner_keys = Vec::<InitializedField>::default();
+				let value = build_struct_json_patch!($type, inner_keys @ { $($body)* });
+				for i in inner_keys.iter_mut() {
+					i.add_prefix(stringify!($key));
+				};
+				$all_keys.extend(inner_keys);
+				value
+			},
+			.. build_struct_json_patch!($struct_type, $all_keys @ $($tail)*)
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: 9fd2dbd36c2

File: substrate/frame/support/src/generate_genesis_config.rs [hunk 19]
```diff
@@ -0,0 +1,875 @@
...
+	) => {
+		{
+			use $crate::generate_genesis_config::{InitializedField, retain_initialized_fields};
+			extern crate alloc;
+			use alloc::{string::ToString, vec::Vec };
+			let mut keys : Vec<InitializedField> = vec![];
+			#[allow(clippy::needless_update)]
+			let struct_instance = build_struct_json_patch!($struct_type, keys @  { $($tail)* });
+			let mut json_value =
+				serde_json::to_value(struct_instance).expect("serialization to json should work. qed");
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/frame/support/src/generate_genesis_config.rs]: done in: 5316807217a

@bkchr you may want to cross check this little commit (added `serde_json` to `frame_support::__private`).

File: polkadot/runtime/westend/src/genesis_config_presets.rs [hunk 20]
```diff
@@ -170,7 +171,7 @@ fn westend_testnet_genesis(
 	const ENDOWMENT: u128 = 1_000_000 * WND;
 	const STASH: u128 = 100 * WND;
 
-	let config = RuntimeGenesisConfig {
+	build_struct_json_patch!(RuntimeGenesisConfig {
```
Comment (bkontur) [inline polkadot/runtime/westend/src/genesis_config_presets.rs:174]: @michalkucharczyk Do you plan to clean up all the other testnet runtimes here or follow up?

**Review by bkontur — APPROVED** [review]

File: polkadot/runtime/westend/src/genesis_config_presets.rs (hunk 20 above)
Comment (michalkucharczyk) [reply polkadot/runtime/westend/src/genesis_config_presets.rs:174]: I planned to do this in follow-up.

Comment (michalkucharczyk) [reply polkadot/runtime/westend/src/genesis_config_presets.rs:174]: @bkontur: https://github.com/paritytech/polkadot-sdk/pull/6349

**Pull request merged**